### ISSUE TITLE
Http 400 Preflight validation check error `the operation failed or was cancelled` needs detailed message

### ISSUE THREAD

**Issue by hemarina** [Azure/azure-sdk-for-go#23350]
### Bug Report

- import path of package in question: [preflight validate](https://github.com/Azure/azure-sdk-for-go/blob/651a245f43c78096eb2be62b638888eb4b584932/sdk/resourcemanager/resources/armresources/deployments_client.go#L2351)
- SDK version: `armresources`, `latest`, `1.2.0`
  - Specify the exact commit if possible: https://github.com/Azure/azure-sdk-for-go/commit/b3194dd32c895011eee5465ab41d338c67455ce1#diff-201f295ddf411fc3a73ca555f5a3da5f3796d1dff4075ec894d07292bc79fc1b
- output of `go version`: `go1.22.1 windows/amd64`

I'm testing [preflight validation function](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armresources#DeploymentsClient.BeginValidate) with an unsupported location (e.g australiacentral) for [todo-nodejs-mongo-aks](https://github.com/Azure-Samples/todo-nodejs-mongo-aks). The error message is [`the operation failed or was cancelled`](https://github.com/Azure/azure-sdk-for-go/blob/82cae1b0c132d35892636ddd10006092ec0fe2eb/sdk/azcore/runtime/poller.go#L88). The error message should provide a detailed message on why it is failing. 

I debugged and this is the error message that got skipped: `The template deployment '****' is not valid according to the validation procedure. The tracking id is '****'. See inner errors for details. Preflight validation check for resource(s) for container service *** in resource group ****** failed. Message: Virtual Machine size: '' is not supported for subscription ***** in location 'australiacentral'. Please refer to aka.ms/aks/vm-size-selector to find supported VM sizes in location 'australiacentral'.` This message is necessary for users to debug and should be provided in error message.

**Comment by tadelesh** [comment]
could you provide your go code of using the `BeginValdate` method?

**Comment by tadelesh** [comment]
refer the following code to deal with error code and error message
```
var rawResponse *http.Response
ctx := context.TODO() // your context
ctxWithResp := runtime.WithCaptureResponse(ctx, &rawResponse)
resp, err := resourceGroupsClient.CreateOrUpdate(ctxWithResp, resourceGroupName, resourceGroupParameters, nil)
if err != nil {
    // with error, you can get RawResponse from context
    log.Printf("Status code: %d", rawResponse.StatusCode)
    var respErr *azcore.ResponseError
    if errors.As(err, &respErr) {
        // with error, you can also get RawResponse from error
        log.Fatalf("Status code: %d", respErr.RawResponse.StatusCode)
    } else {
        log.Fatalf("Other error: %+v", err)
    }
}
// without error, you can get RawResponse from context
log.Printf("Status code: %d", rawResponse.StatusCode)
```

**Comment by weikanglim** [comment]
I can add some debugging notes here since I was fortunate enough to be present with @hemarina at the time, but I do believe this is a bug in the Swagger spec that is used to generate `deployment_clients.go`.

Here's the execution flow we observed:
1. Client calls [BeginValidate](https://github.com/Azure/azure-sdk-for-go/blob/790270459f3d16ed74e15005b4936716d72667dc/sdk/resourcemanager/resources/armresources/deployments_client.go#L2316) in deployment_clients.go.
2. ARM returns 400 Bad Request
3. [validate](https://github.com/Azure/azure-sdk-for-go/blob/790270459f3d16ed74e15005b4936716d72667dc/sdk/resourcemanager/resources/armresources/deployments_client.go#L2351) checks for Bad Request, but considers it a success state, it returns `httpResp, nil` instead of `nil, err`.
4.  Execution continues until [NewPoller](https://github.com/Azure/azure-sdk-for-go/blob/790270459f3d16ed74e15005b4936716d72667dc/sdk/azcore/runtime/poller.go#L87), where it's rejected by the validation check.

In summary, the bug observed by the user is:

- When ARM returns 400 bad request for `deployments/validate`, the error `the operation failed or was cancelled` is returned instead of the expected `azcore.ResponseError`.

We are also happy to provide a minimal repro setup, but it boils down to calling deploymentsClient.BeginValidate with parameters that will result in an ARM 400 response.

**Comment by tadelesh** [comment]
@weikanglim thanks for the detail. if possible, could you provide the detail logs with this [instruction](https://github.com/Azure/azure-sdk-for-go/blob/main/documentation/new-version-guideline.md#logging). i want to confirm the service return 400 with the lro init call.

**Comment by rajeshkamal5050** [comment]
@hemarina can you enable logging and share?

**Comment by tadelesh** [comment]
@jhendrixMSFT For LRO with `400` response definition, could it be the initial request response result? (see spec [here](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/resources/resource-manager/Microsoft.Resources/stable/2024-07-01/resources.json#L283))
@hemarina Before we confirm if it is a service issue, I think you could also try to get the error info with my comment [code](https://github.com/Azure/azure-sdk-for-go/issues/23350#issuecomment-2301639365) with the first call of LRO, it should also work for LRO.

**Comment by jhendrixMSFT** [comment]
Including a `400` in the responses indicates it should be treated as a success response. However, if you look at the schema for the `400` it's an error type. The service team either needs to remove the `400` or add [`x-ms-error-response`](https://github.com/Azure/autorest/tree/main/docs/extensions#x-ms-error-response) to it.

**Comment by tadelesh** [comment]
Got it. Thanks Joel. Let me add service attention label.

**Comment by tadelesh** [comment]
Our SDK is auto-generated from service's spec. @armleads-azure could you help to prioritize?

**Comment by jhendrixMSFT** [comment]
@tadelesh `armresources` is on an old API version [`2024-04-01`](https://github.com/Alancere/azure-sdk-for-go/blob/main/sdk/resourcemanager/resources/armresources/deployments_client.go#L2310). Can we get this updated? It's been fixed in later API versions (2025-03-01 for sure, I didn't check earlier).

**Comment by jhendrixMSFT** [comment]
Looks like the `armresources` SDK has been split into several smaller SDKs. I think the functionality is now in the [`armdeploymentstacks`](https://github.com/Azure/azure-sdk-for-go/tree/main/sdk/resourcemanager/resources/armdeploymentstacks) SDK.

**Comment by hemarina** [comment]
@jhendrixMSFT Thank you for your kind message. I confirmed that latest armdeploymentstacks works for this bug. For standard deployment like [armresources](https://pkg.go.dev/github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/resources/armresources#DeploymentsClient.BeginValidate), the bug still exists. May I ask what's the plan for standard deployment's BeginValidate function? In [changelogs](https://github.com/Azure/azure-sdk-for-go/blob/main/sdk/resourcemanager/resources/armresources/CHANGELOG.md), BeginValidate is removed. Does this mean that preflight validate api will only be supported in deployment stack?

**Comment by jhendrixMSFT** [comment]
> Does this mean that preflight validate api will only be supported in deployment stack?

That's correct. The service team split `armresources` into several SDKs.

Generally speaking, only the latest major version of an SDK is supported. We won't be updating the v1 or v2 SDKs to fix this.

**Comment by andyliuliming** [comment]
https://github.com/Azure/azure-sdk-for-go/blob/0767c4b794572397c9ba7146b7ca2e703caddeba/sdk/resourcemanager/resources/armdeploymentstacks/client.go#L1176

looks like when the response is 400, that error(without details) still returned?

**Comment by jhendrixMSFT** [comment]
There appear to be two problems here. The first is that the service (still) models a HTTP 400 as a successful response.

https://github.com/Azure/azure-rest-api-specs/blob/main/specification/resources/resource-manager/Microsoft.Resources/deploymentStacks/stable/2024-03-01/deploymentStacks.json#L387-L392

What this means is that SDKs produced by our code generators will treat the 400 as a _successful response_ and _not_ return an error. So, removing the 400 from the swagger should fix the immediate problem.

The second (larger) problem is that it appears that the service is returning a 400 as the initial response to initiate a long-running operation. This is forbidden by the [Azure API Guidelines](https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#lro-action-on-a-resource-pattern) and should be fixed by the service in a new API version.

**Comment by jhendrixMSFT** [comment]
Added a workaround in `azcore` which should resolve the immediate issue. https://github.com/Azure/azure-sdk-for-go/pull/25206

However, the service definition and behavior is still incorrect and should be fixed.
